File: golang/array_string/string_compression.go
package main

import "strconv"

func compress(chars []byte) int {
	write := 0

	for read := 0; read < len(chars); {
		char := chars[read]

		count := 0

		for read < len(chars) && chars[read] == char {
			read++

			count++
		}

		chars[write] = char

		write++

		if count > 1 {
			for _, c := range strconv.Itoa(count) {
				chars[write] = byte(c)

				write++
			}
		}
	}

	return write
}

File: java/src/main/java/com/sanedge/interview/array_string/StringCompression.java
package com.sanedge.interview.array_string;

public class StringCompression {

    public int compress(char[] chars) {
        int write = 0;

        for (int read = 0; read < chars.length;) {
            char currentChar = chars[read];
            int count = 0;

            while (read < chars.length && chars[read] == currentChar) {
                read++;
                count++;
            }

            chars[write] = currentChar;
            write++;

            if (count > 1) {
                for (char c : Integer.toString(count).toCharArray()) {
                    chars[write] = c;
                    write++;
                }
            }
        }

        return write;
    }
}
